File: projects/social_media_dashboard/app/src/main/java/com/vishnu/socialdashboard/data/repository/UserRepository.java
package com.vishnu.socialdashboard.data.repository;

import com.vishnu.socialdashboard.data.api.ApiClient;
import com.vishnu.socialdashboard.data.api.SocialApiService;
import com.vishnu.socialdashboard.data.model.User;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class UserRepository {

    private final SocialApiService apiService;

    public UserRepository() {
        this.apiService = ApiClient.getInstance().getApiService();
    }

    public interface UserCallback {
        void onSuccess(User user);
        void onError(String message);
    }

    public void getUser(int userId, UserCallback callback) {
        apiService.getUser(userId).enqueue(new Callback<User>() {
            @Override
            public void onResponse(Call<User> call, Response<User> response) {
                if (response.isSuccessful() && response.body() != null) {
                    callback.onSuccess(response.body());
                } else {
                    callback.onError("Failed to load user");
                }
            }

            @Override
            public void onFailure(Call<User> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }
}

File: projects/chat_app/app/src/main/java/com/vishnu/chatapp/data/repository/UserRepository.kt
package com.vishnu.chatapp.data.repository

import com.vishnu.chatapp.data.model.User
import com.vishnu.chatapp.data.remote.FirestoreService
import kotlinx.coroutines.flow.Flow

class UserRepository(
    private val firestoreService: FirestoreService = FirestoreService()
) {

    fun getUsers(): Flow<List<User>> {
        return firestoreService.getUsers()
    }

    suspend fun getUser(userId: String): User? {
        return firestoreService.getUser(userId)
    }

    suspend fun updateUserStatus(userId: String, status: String) {
        firestoreService.updateUserStatus(userId, status)
    }

    suspend fun updateUser(user: User) {
        firestoreService.createOrUpdateUser(user)
    }
}
